### PR TITLE
Add support for evaluating the flipsign function for complex numbers 

### LINKED ISSUE #102: RFC: complex number flipsign
## Checklist

> Please ensure the following tasks are completed before filing an issue.

* [x] Read and understood the [Code of Conduct][code-of-conduct].
* [x] Searched for existing issues and pull requests.
* [x] If this is a general question, searched the [FAQ][faq] for an existing answer.
* [x] If this is a feature request, the issue name begins with `RFC: `.

## Description

> Description of the issue (or feature request).

Add support for returning a complex number which has the same magnitude as `z` and is `-z` if `y` is negative. Equivalent to `flipsign` for reals.

Package: `@stdlib/math/base/complex/flipsign`
alias: `cflipsign`

Behavior:

``` javascript
var v = cflipsign( re, im, +y );
// returns [ re, im ]

v = cflipsign( re, im, -y );
// returns [ -re, -im ]
```

## Related Issues

> Does this issue (or feature request) have any related issues?

None.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this issue (or feature request)? This may include screenshots, references, stack traces, sample output, and/or implementation notes.

No.

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md
[faq]: https://github.com/stdlib-js/stdlib/blob/develop/FAQ.md

### PR DESCRIPTION
**Pull request by ShraddheyaS**
Resolves #102.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [X] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [X] Read and understand the [Code of Conduct][code-of-conduct].
-   [X] Read and understood the [licensing terms][license].
-   [X] Searched for existing issues and pull requests **before** submitting this pull request.
-   [X] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [X] Rebased onto latest `develop`.
-   [X] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds support for evaluating the flipsign function for complex numbers 

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #102.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/benchmark.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	var im;
+	var y;
+	var i;
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		re = ( randu()*1000.0 ) - 500.0;
+		im = ( randu()*1000.0 ) - 500.0;
+		y = ( randu()*1000.0 ) - 500.0;
+		ans = ( y < 0 || isNegativeZero( y ) ) ? [ -re, -im ] : [ re, im ];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/benchmark.js]: You probably don't need to be so precise here. The likelihood of `-0` is essentially nil. In which case, you can remove the `isNegativeZero` check.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/benchmark.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	var i;
+
+	ans = new Array( 2 );
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		re = ( randu()*1000.0 ) - 500.0;
+		im = ( randu()*1000.0 ) - 500.0;
+		y = ( randu()*1000.0 ) - 500.0;
+		ans[ 0 ] = ( y < 0 || isNegativeZero( y ) ) ? -re : re;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/benchmark.js]: Same comment as above. And further, these checks can be consolidated into a single conditional.

```javascript
if ( y < 0 ) {
    re *= -1;
    im *= -1;
}
```

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/benchmark.js]: In short, one of the golden rules when writing benchmarks is "Don't do anything which **needlessly** slows down benchmarked code." Unnecessary branching violates this rule.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/julia/benchmark.jl [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+* The elapsed time is in seconds.
+
+# Examples
+
+``` julia
+julia> out = benchmark();
+```
+"""
+function benchmark()
+	t = BenchmarkTools.@benchmark sign( Complex128( (rand()*1000.0) - 500.0, (rand()*1000.0) - 500.0 ) ) samples=1e6
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/benchmark/julia/benchmark.jl]: s/sign/flipsign/

And an additional argument needs to be provided to the `flipsign` function invocation here.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        Output array.
+
+    re: number
+        Real component.
+
+    im: number
+        Imaginary component.
+    
+    y: number
+        Number from which to derive the sign
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt]: Missing the ending period.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt [hunk 5]
```diff
@@ -0,0 +1,38 @@
+
+{{alias}}( [out,] re, im )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt]: The function signature is incorrect.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    Returns a complex number which has the same magnitude as `z` and is `-z` if `y` is negative.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt]: Does the description

> Returns a complex number with the same magnitude as `z` and the sign of `y*z`.

not still work?

Something about the description seems a bit awkward to me.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt]: Also, lines in REPL text files should be wrapped at 80 characters. For more details on writing REPL text files, see the [guide](https://github.com/stdlib-js/stdlib/blob/develop/docs/repl_text.md).

File: lib/node_modules/@stdlib/math/base/complex/flipsign/lib/cflipsign.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+* var v = cflipsign( out, -4.2, 5.5, 55 );
+* // returns [ -4.2, 5.5 ]
+*
+* var bool = ( v === out );
+* // returns true
+*
+* @example
+* var out = new Array( 2 );
+* var v = cflipsign( out, 0.0, 0.0, -8 );
+* // returns [ 0.0, 0.0 ]
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/lib/cflipsign.js]: returns positive or negative zeros?

File: lib/node_modules/@stdlib/math/base/complex/flipsign/lib/index.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+* @module @stdlib/math/base/complex/flipsign
+*
+* @example
+* var cflipsign = require( '@stdlib/math/base/complex/flipsign' );
+*
+* var v = cflipsign( -4.2, 5.5, 5 );
+* // returns [ -4.2, 5.5 ]
+*
+* v = cflipsign( 0.0, 0.0, -99 );
+* // returns [ 0.0, 0.0 ]
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/lib/index.js]: returns positive or negative zeros?

File: lib/node_modules/@stdlib/math/base/complex/flipsign/lib/main.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+*
+* var v = cflipsign( out, -4.2, 5.5, 55 );
+* // returns [ -4.2, 5.5 ]
+*
+* var bool = ( v === out );
+* // returns true
+*
+* @example
+* var v = cflipsign( 0.0, 0.0, -8 );
+* // returns [ 0.0, 0.0 ]
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/lib/main.js]: returns positive or negative zeros?

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+		'1': 5.5
+	};
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+	t.strictEqual( actual, out, 'returns output value' );
+
+	t.end();
+});
+
+tape( 'the function does not flip sign if provided NaN `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: "the function does not flip the sign of real and imaginary input arguments if `y` is `NaN`"

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+	expected = [ -4.2, 5.5 ];
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+	t.strictEqual( actual, out, 'returns output value' );
+
+	t.end();
+});
+
+tape( 'the function does not flip sign if provided `+infinity` as `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: "the function does not flip the sign of real and imaginary input arguments if `y` is `+infinity`"

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+
+	expected = [ -4.2, 5.5 ];
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+	t.strictEqual( actual, out, 'returns output value' );
+
+	t.end();
+});
+
+tape( 'the function flips sign if provided `-infinity` as `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: "the function flips the sign of real and imaginary input arguments if `y` is `-infinity`"

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+
+	expected = [ 4.2, -5.5 ];
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+	t.strictEqual( actual, out, 'returns output value' );
+
+	t.end();
+});
+
+tape( 'the function returns a `NaN` if provided a `NaN`, positive `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: Tests L188, 195, and L202 can be consolidated into a single test which includes different `y` values. Splitting into separate tests is not necessary and does not add any conceptual clarity.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+});
+
+tape( 'the function returns a `NaN` if provided a `NaN`, NaN `y`', function test( t ) {
+	var val = cflipsign( NaN, NaN, NaN );
+	t.strictEqual( isnan( val[ 0 ] ), true, 'returns expected value' );
+	t.strictEqual( isnan( val[ 1 ] ), true, 'returns expected value' );
+	t.end();
+});
+
+tape( 'the function returns `+0` if provided `+0`, positive `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: "the function returns `+0` if provided `+0` and a positive `y`"

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+	var val = cflipsign( +0.0, +0.0, 4 );
+	t.strictEqual( isPositiveZero( val[ 0 ] ), true, 'returns expected value' );
+	t.strictEqual( isPositiveZero( val[ 1 ] ), true, 'returns expected value' );
+	t.end();
+});
+
+tape( 'the function returns `-0` if provided `-0`, positive `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: "the function returns `-0` if provided `-0` and a positive `y`"

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+	var val = cflipsign( -0.0, -0.0, 55 );
+	t.strictEqual( isNegativeZero( val[ 0 ] ), true, 'returns expected value' );
+	t.strictEqual( isNegativeZero( val[ 1 ] ), true, 'returns expected value' );
+	t.end();
+});
+
+tape( 'the function returns `+0` if provided `-0`, negative `y`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: Similar change as above. Basically, the change is to make the description a more grammatically correct sentence.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js]: This comment also applies to all the descriptions which follow.

**Review by kgryte — CHANGES_REQUESTED** [review]
@ShraddheyaS Thanks for working on this! For the most part, everything looks good. Just a few questions and some minor clean-up.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt [hunk 17]
```diff
@@ -0,0 +1,39 @@
+
+{{alias}}( [out,] re, im, y )
+    Returns a complex number with the same magnitude as `z` and
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt:3]: This same description can be applied elsewhere.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt:3]: ...that is if you believe that the description is appropriate. I suppose it depends on whether it is fair to say the "sign" of a complex number. Do you have strong opinions here?

Comment (ShraddheyaS) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt:3]: I think that the `complex/signum` might serve as the "sign" in this case. I can't seem to think of a corner-case.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt:3]: In which case, I think it should be fine to use a description consistent with `special/flipsign`.

Comment (ShraddheyaS) [reply lib/node_modules/@stdlib/math/base/complex/flipsign/docs/repl.txt:3]: So yes, it does make sense if we assume the sign of a complex number to be the signum function.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/README.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+
+## Usage
+
+```javascript
+var cflipsign = require( '@stdlib/math/base/complex/flipsign' );
+```
+
+#### cflipsign( \[out,] re, im )
+
+Returns a complex number which has the same magnitude as `z` and is `-z` if `y` is negative.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/README.md]: One more description to change.

File: lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+		'1': 5.5
+	};
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+	t.strictEqual( actual, out, 'returns output value' );
+
+	t.end();
+});
+
+tape( 'the function does not flip the sign of real and imaginary input arguments if `y` is `NaN`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/complex/flipsign/test/test.js:137]: One final set of tests which is not included is what happens when `y` is `+-0`. :)

In other words, we don't have any tests which explicitly exercise the `isNegativeZero` branch.

**Review by kgryte — CHANGES_REQUESTED** [review]
Apart from the tests when `y` is `+-0`, this PR is ready for merge! Excellent work!

**Review by kgryte — APPROVED** [review]

**Pull request merged**